rfcs/extended_goal_state_functionality.md

```diff
@@ -0,0 +1,72 @@
+
+# Implementation
+
+![Information Flow](extended_goal_state_functionality/architecture_overview.png)
+**Proposed Routing Pathways**. Learning module (LM) and sensor module (SM) derived goal states will be routed to the `GoalStateSelector`, and the `GoalStateSelector`'s output will be routed to the motor system.
+
+The `GoalStateSelector` will receive any number of goal states derived from learning and sensor modules, and it will produce at most one goal state for consumption by the motor system. There are many potential strategies for choosing a goal state, but the simplest initial approach will be to select the goal state with the largest `confidence` value. In this way, it will act exactly as the `Monty` class currently does. More complex methods are outside the scope of this RFC.
+
+Many other Monty components will require modification. Below is a non-exhaustive list of anticipated changes.
+
+### `SensorModule`
+ - Will have an optional `gsg` attribute (i.e., it will be a `GoalStateGenerator` or `None`).
+ - Will call its `gsg`'s `step_gsg` method before exiting its own `step` method (if `gsg` is not `None`).
+ - Will have a `propose_goal_state` method that returns any number of goal states. Note that this differs from the learning module's `propose_goal_state` method which returns at most 1 goal state. Returing many goal states will enable planned work on salience-based targeting methods.
+
+All of the above changes are meant to mirror existing learning-module/GSG behavior.
+
+### `Monty`
+ - Will have a `gss` attribute to hold a `GoalStateSelector`.
+ - Existing machinery that collects goal states, performs selection, and updates the motor system will be replaced. More specifically, `Monty` will collect goal states from both learning modules and sensor modules, but it will defer goal state selection to its `gss` object. The output of the `gss` object will then be used to update the motor system. See the for-loop associated with `best_goal_confidence` in `_pass_infos_to_motor_system` for how this is currently implemented.
+ - Optionally, `Monty` may be equipped with additional `lm_to_gss` and `sm_to_gss` connectivity matrices to support more complex routing pathways, but this is not an immediate need.
+
+### `MotorSystem`
+ - Will be initialized with any number of policies. As before, these will all be model-free.
+ - Will perform policy/action selection based on a given goal state. Note that a goal state may not always be set, in which case it is expected that an appropriate model-free policy will be selected to produce an action. For example, when goal-states from `GetGoodView` are not used, the motor system can randomly saccade.
+
+### `GoalState`
+ - May be augmented with metadata indicating the target of the goal, such as the agent's position or a sensor's sensed location.
```

Comment:
Another potential extention of the goal state we talked about is that it may specify an area instead of a point in space. This could be interpreted as "observe somewhere in this rough area" (uncertainty) or "observe this area" (implicitly telling agent to move closer/further away from object)